### ISSUE TITLE
Instrumentation is not thread safe

### ISSUE THREAD

**Issue by FFMG** [TheCherno/Hazel#176]
Not sure if this is your intention, but your Instrumentation is not thread safe.

The way I solved it is by adding a lock around `WriteProfile`.

```cpp
// ...
std::mutex _mtx;  

// ...
void WriteProfile(const ProfileResult& result)
{
  _mtx.lock();
  try
  {
    // ...
  }
  catch( ... )
  {
  }
  _mtx.unlock();
}
```
Let me know if you would like me to submit a PR but I suspect you already have your own approach to this.

Of course the price of the lock is kind of rendering the profiling pointless in multi-threading cases.

**Comment by Emilundpixeln** [comment]
I would recommend to use something like `std::stringstream` and write to that first and then just do one mutex protected write to the file to minimize the time the mutex is locked. Also I think `std::ofstream` doesn't throw exceptions by default (https://stackoverflow.com/questions/10337915/ofstream-exception-handling).

**Comment by FFMG** [comment]
Yes, I did something similar in my own branch.
In fact, I even cache 'x' number of measurements before writing them to disk.

If there is a crash the file is not closed properly anyway, (and I don't use profiling for crash investigations anyway).
I prefer speed so my profiling is as accurate as possible.

**Comment by freeman40** [comment]
In #PR179,  I've used lock_guard rather than have that try..catch in there.
BeginSession() and EndSession() also need the mutex to be locked.  (e.g. what happens if one thread tries to end session and begin another one,  while some other thread is WriteProfile()'ing.

**Comment by freeman40** [comment]
@FFMG That is a good point about potential waits for the lock messing up multi-threaded profiling results anyway.   (but with the lock, at least the result file will be valid json)

**Comment by freeman40** [comment]
> 
> 
> I would recommend to use something like `std::stringstream` and write to that first and then just do one mutex protected write to the file to minimize the time the mutex is locked....

Does anyone know if a single call to the insertion operator is atomic?  If it were, then following @Emilundpixeln suggestion here, would mean that the mutex isn't needed at all.

The internet suggests that ostream << is atomic on MSVC++, but I couldn't find anything about other compilers.  In c++20 there is osyncstream

**Comment by freeman40** [comment]
tldr:  I we do @Emilundpixeln suggestion, then the mutex can (very likely) be removed.

Details:
I think if WriteProfile() is limited to a single insertion into the output stream, then the mutex is not needed.  It appears that on the following compilers, the insertion operator is atomic:
- Visual Studio 2019 16.3.10
- gcc 9.2.0
- clang 9.0.0

I had to try quite hard to get garbled output from a multi-threaded test program.  FWIW here is the test code:
This will produce garbled output on all of the compilers mentioned above - but the garbling only ever occurs between invocations of the insertion operator.  Each individual << appears to be atomic.

If you change Worker() to only do a single <<,  then I couldn't make it garble the output (and I tried several different things).

`#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <thread>
#include <vector>

void Worker(std::ofstream& os, int delay) {
   std::this_thread::sleep_for(std::chrono::milliseconds(delay));
   if (os.is_open()) {
      os << "_This is quite a long string, could insertion operator get pre-empted_";
      os << "**Timestamp:" << std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count() << "**";
      std::this_thread::sleep_for(std::chrono::milliseconds(1));
      os << "**Thread:" << delay / 100 << "**";
   }
}

int main() {
   std::filesystem::path filepath = std::filesystem::temp_directory_path() / "a.out";
   std::ofstream os(filepath.string());

   std::vector<std::unique_ptr<std::thread>> threads;

   for (int i = 0; i < 16; ++i) {
      threads.emplace_back(std::make_unique<std::thread>(Worker, std::ref(os), (16 - i) * 100));
      std::cout << "spawned thread " << i << std::endl;
      std::this_thread::sleep_for(std::chrono::milliseconds(99));
   }

   for (auto& thread : threads) {
      thread->join();
   }

   os.close();

   std::cout << std::ifstream(filepath.string()).rdbuf() << std::endl;

}
`

**Comment by Emilundpixeln** [comment]
I have looked into the implementation of `std::fstream` for windows and for writing to a file it uses `fwrite` which is thread-safe (https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fwrite?view=vs-2019).
I don't know how it is on other platforms.

**Comment by Puddlestomper** [comment]
As far as I know the C++ standard provides no guarantee that the operation is thread safe so we really shouldn't be relying on that. Especially since this is in the platform-agnostic part of Hazel.

**Comment by ReDucTor** [comment]
http://eel.is/c++draft/iostreams.requirements

> Concurrent access to a stream object ([string.streams], [file.streams]), stream buffer object ([stream.buffers]), or C Library stream ([c.files]) by multiple threads may result in a data race ([intro.multithread]) unless otherwise specified ([iostream.objects]). [ Note: Data races result in undefined behavior ([intro.multithread]). — end note ]

Ideally a profiler shouldn't be writing like this during its annotations it should be left to be done at regular intervals in a background thread, file I/O is extremely expensive and as is locking, you don't want to be doing either regularly.

For solving this I recommend forming a list (I don't mean an std::list) of the items to be written in a thread safe manor then processing them at a regular interval in a background thread

**Comment by freeman40** [comment]
@Puddlestomper You are absolutely correct that generally one shouldn't rely on a an implementation detail of one particular library that may not be true on other platforms.
In this case it may be justified because:
- With MSVC on windows we know from looking through the std library source code that a single call to ostream << is indeed atomic.   MSVC on windows is currently the only "supported" platform for Hazel.  On this platform, adding the mutex in WriteProfile() is achieving nothing except for extra work.
- This is also "developer only" type code.  Use of the profile results is likely to involve adding further instrumentation, recompiling, and an intimate knowledge of how the profiling works - it would not be much of a stretch for that developer (on a non-MSVC platform) to add in a lock if they discovered their profile results file is jumbled.

Perhaps a compromise?  No locking in WriteProfile() if compiling on windows with MSVC,  otherwise lock?  What do you reckon?

@ReDucTor  Fair enough and good suggestion.  I'm wondering though whether the added complexity in the profiling system is warranted.  Hazel is about learning game engine architecture, and I feel like we're getting a bit bogged down coming up with a gold-plated instrumentation profiler (ok, maybe not "gold" exactly, but you get my point).  I think that something simple that gives "broad brush" indications of bottlenecks (and the output isn't garbled) is good enough at this stage.

**Comment by freeman40** [comment]
Instead of rolling our own profiling code, we could instead use a third-party profiler.
[Tracy Profiler](https://bitbucket.org/wolfpld/tracy/src/master/) looks quite good.  It is cross platform, and low overhead.  It would not even be necessary to re-instrument the Hazel code - just replace the body of HZ_PROFILE_FUNCTION() and HZ_PROFILE_SCOPE() with the equivalent tracy profiler incantations.

**Comment by Lecrapouille** [comment]
@freeman40 you are right it's better to use an already proof-crash tool than to reinvent the wheel (especially when it's so easy to call gprof) but this is good for a general overview of your project. I think it's always good to have a 100-line header-only utility tool than can quickly be inserted in your project and help you graphically to debug your 3-4 specific functions than linking your project against a huge library and read 50 pages of pdf to make it works and have a trace about 1000 functions you do not care for at final to obtain a poor human-friendly report. For example, this helped me to understand in 5 min that my setup() function was too long because SOIL was not very fast. Of course, with tracy and gprof I would have obtained the same result. In my opinion `#if HZ_PROFILE` in profiler.h and letting `HZ_PROFILE_FUNCTION` everywhere in the client code is not a good idea because the goal of this tool is to help to debug so macros shall always be accessible and tracing macros inside client code not committed.
